File: routes/word_editor.py
from flask import Blueprint, request, jsonify, send_file
from docx import Document
from docx.shared import Pt, RGBColor, Inches
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.oxml import OxmlElement
from docx.oxml.ns import qn
from reportlab.lib.pagesizes import letter
from reportlab.lib.enums import TA_LEFT, TA_CENTER, TA_RIGHT, TA_JUSTIFY
from reportlab.lib.colors import HexColor
from reportlab.lib.styles import ParagraphStyle
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer
from io import BytesIO
from datetime import datetime

word_editor_bp = Blueprint('word_editor', __name__)

# Simple in-memory document storage for demo
documents = []
doc_id_counter = 1

DEFAULT_FORMATTING = {
    'fontFamily': 'Arial',
    'fontSize': 12,
    'fontColor': '#000000',
    'bold': False,
    'italic': False,
    'underline': False,
    'margins': {
        'top': 1,
        'bottom': 1,
        'left': 1,
        'right': 1
    },
    'alignment': 'left',
    'lineSpacing': 1.5,
    'pageNumbers': False
}


def merge_formatting(existing, new_values):
    formatting = existing.copy()
    if not new_values:
        return formatting

    for key, value in new_values.items():
        if key == 'margins' and isinstance(value, dict):
            merged_margins = formatting.get('margins', DEFAULT_FORMATTING['margins']).copy()
            merged_margins.update({k: float(v) for k, v in value.items() if v is not None})
            formatting['margins'] = merged_margins
        else:
            formatting[key] = value
    return formatting


def inches(value, default=1.0):
    try:
        return Inches(float(value))
    except (TypeError, ValueError):
        return Inches(default)


def parse_color(color_str):
    if not color_str:
        return None
    color = color_str.lstrip('#')
    if len(color) != 6:
        return None
    try:
        return RGBColor.from_string(color.upper())
    except ValueError:
        return None


def apply_run_formatting(run, formatting):
    font = run.font
    font.name = formatting.get('fontFamily') or DEFAULT_FORMATTING['fontFamily']

    font_size = formatting.get('fontSize')
    try:
        font.size = Pt(float(font_size))
    except (TypeError, ValueError):
        font.size = Pt(DEFAULT_FORMATTING['fontSize'])

    color = parse_color(formatting.get('fontColor'))
    if color:
        font.color.rgb = color

    run.bold = bool(formatting.get('bold'))
    run.italic = bool(formatting.get('italic'))
    run.underline = bool(formatting.get('underline'))


def apply_paragraph_formatting(paragraph, formatting):
    pf = paragraph.paragraph_format
    alignment_map = {
        'left': WD_ALIGN_PARAGRAPH.LEFT,
        'center': WD_ALIGN_PARAGRAPH.CENTER,
        'right': WD_ALIGN_PARAGRAPH.RIGHT,
        'justify': WD_ALIGN_PARAGRAPH.JUSTIFY
    }
    alignment = formatting.get('alignment', DEFAULT_FORMATTING['alignment'])
    paragraph.alignment = alignment_map.get(alignment, WD_ALIGN_PARAGRAPH.LEFT)

    line_spacing = formatting.get('lineSpacing', DEFAULT_FORMATTING['lineSpacing'])
    try:
        pf.line_spacing = float(line_spacing)
    except (TypeError, ValueError):
        pf.line_spacing = DEFAULT_FORMATTING['lineSpacing']


def add_page_numbers(document):
    section = document.sections[0]
    footer = section.footer
    paragraph = footer.paragraphs[0] if footer.paragraphs else footer.add_paragraph()
    paragraph.alignment = WD_ALIGN_PARAGRAPH.CENTER
    run = paragraph.add_run()

    fld_char_begin = OxmlElement('w:fldChar')
    fld_char_begin.set(qn('w:fldCharType'), 'begin')
    run._r.append(fld_char_begin)

    instr_text = OxmlElement('w:instrText')
    instr_text.text = 'PAGE'
    run._r.append(instr_text)

    fld_char_end = OxmlElement('w:fldChar')
    fld_char_end.set(qn('w:fldCharType'), 'end')
    run._r.append(fld_char_end)


def build_docx_document(doc):
    document = Document()
    formatting = doc.get('formatting', DEFAULT_FORMATTING)

    # Apply section margins
    section = document.sections[0]
    margins = formatting.get('margins', DEFAULT_FORMATTING['margins'])
    section.top_margin = inches(margins.get('top', 1))
    section.bottom_margin = inches(margins.get('bottom', 1))
    section.left_margin = inches(margins.get('left', 1))
    section.right_margin = inches(margins.get('right', 1))

    # Add content
    paragraphs = [p for p in doc.get('content', '').split('\n') if p is not None]
    for para in paragraphs:
        text = para.rstrip()
        paragraph = document.add_paragraph()
        if text.strip():
            run = paragraph.add_run(text)
        else:
            run = paragraph.add_run(' ')
        apply_run_formatting(run, formatting)
        apply_paragraph_formatting(paragraph, formatting)

    if formatting.get('pageNumbers'):
        add_page_numbers(document)

    buffer = BytesIO()
    document.save(buffer)
    buffer.seek(0)
    return buffer


def build_pdf_document(doc):
    formatting = doc.get('formatting', DEFAULT_FORMATTING)
    buffer = BytesIO()

    margins = formatting.get('margins', DEFAULT_FORMATTING['margins'])
    left_margin = float(margins.get('left', 1)) * 72
    right_margin = float(margins.get('right', 1)) * 72
    top_margin = float(margins.get('top', 1)) * 72
    bottom_margin = float(margins.get('bottom', 1)) * 72

    pdf_doc = SimpleDocTemplate(
        buffer,
        pagesize=letter,
        leftMargin=left_margin,
        rightMargin=right_margin,
        topMargin=top_margin,
        bottomMargin=bottom_margin
    )

    font_family_map = {
        'Arial': 'Helvetica',
        'Calibri': 'Helvetica',
        'Georgia': 'Times-Roman',
        'Times New Roman': 'Times-Roman'
    }

    base_font = font_family_map.get(formatting.get('fontFamily'), 'Helvetica')
    font_variant = base_font

    bold = formatting.get('bold')
    italic = formatting.get('italic')

    if base_font == 'Times-Roman':
        if bold and italic:
            font_variant = 'Times-BoldItalic'
        elif bold:
            font_variant = 'Times-Bold'
        elif italic:
            font_variant = 'Times-Italic'
    else:
        # Helvetica/Courier naming
        if bold and italic:
            font_variant = f'{base_font}-BoldOblique'
        elif bold:
            font_variant = f'{base_font}-Bold'
        elif italic:
            font_variant = f'{base_font}-Oblique'

    alignment_map = {
        'left': TA_LEFT,
        'center': TA_CENTER,
        'right': TA_RIGHT,
        'justify': TA_JUSTIFY
    }

    line_spacing = float(formatting.get('lineSpacing', DEFAULT_FORMATTING['lineSpacing']))
    font_size = float(formatting.get('fontSize', DEFAULT_FORMATTING['fontSize']))

    style = ParagraphStyle(
        'Custom',
        fontName=font_variant,
        fontSize=font_size,
        leading=font_size * line_spacing,
        alignment=alignment_map.get(formatting.get('alignment', 'left'), TA_LEFT),
        textColor=HexColor(formatting.get('fontColor', '#000000')),
        underline=bool(formatting.get('underline'))
    )

    elements = []
    for para in doc.get('content', '').split('\n'):
        text = para.strip()
        if not text:
            elements.append(Spacer(1, font_size * 0.6))
        else:
            elements.append(Paragraph(text.replace('\n', '<br/>'), style))
            elements.append(Spacer(1, font_size * 0.3))

    pdf_doc.build(elements or [Paragraph(' ', style)])
    buffer.seek(0)
    return buffer

@word_editor_bp.route('/upload-doc', methods=['POST'])
def upload_doc():
    try:
        print("Upload request received")
        print("Content-Type:", request.headers.get('Content-Type'))
        print("Body:", request.get_json())

        data = request.get_json()
        filename = data.get('filename', 'uploaded_document.docx') if data else 'uploaded_document.docx'
        
        # Generate realistic document content based on filename
        if 'resume' in filename.lower() or 'cv' in filename.lower():
            document_content = """RESUME

John Doe
Software Engineer
Email: john.doe@email.com
Phone: (555) 123-4567

EXPERIENCE
Senior Software Engineer at TechCorp (2020-Present)
• Developed web applications using React and Node.js
• Led a team of 5 developers
• Improved application performance by 40%

Software Engineer at StartupXYZ (2018-2020)
• Built mobile applications using React Native
• Collaborated with design team on user interfaces

EDUCATION
Bachelor of Computer Science
University of Technology (2014-2018)

SKILLS
• JavaScript, Python, Java
• React, Node.js, MongoDB
• Git, Docker, AWS

ACHIEVEMENTS
• Published 3 technical articles
• Speaker at 2 tech conferences
• AWS Certified Solutions Architect"""
        elif 'report' in filename.lower() or 'analysis' in filename.lower():
            document_content = """PROJECT REPORT

Title: Analysis of Student Performance in Online Learning

ABSTRACT
This report analyzes the impact of online learning on student performance during the academic year 2023-2024. The study examines various factors affecting student engagement and academic outcomes.

INTRODUCTION
The transition to online learning has significantly changed educational methodologies. This study examines the effectiveness of digital learning platforms and their impact on student achievement.

METHODOLOGY
Data was collected from 500 students across different departments using:
• Online surveys and questionnaires
• Academic performance records
• Engagement metrics from learning platforms
• Focus group discussions

FINDINGS
1. Student engagement decreased by 15% in online formats
2. Technical difficulties affected 30% of students
3. Self-motivated students performed better in online settings
4. Interactive content improved retention by 25%

CONCLUSION
Online learning presents both challenges and opportunities for educational institutions. Proper implementation and support systems are crucial for success.

RECOMMENDATIONS
• Improve technical infrastructure
• Provide training for both students and faculty
• Develop hybrid learning models
• Implement interactive learning tools"""
        elif 'letter' in filename.lower() or 'application' in filename.lower():
            document_content = f"""APPLICATION LETTER

[Date: {datetime.now().strftime('%Y-%m-%d')}]

Dear Hiring Manager,

I am writing to express my strong interest in the Software Engineer position at your esteemed organization. With my background in computer science and hands-on experience in web development, I am confident that I would be a valuable addition to your team.

QUALIFICATIONS
• Bachelor's degree in Computer Science
• 3+ years of experience in software development
• Proficiency in React, Node.js, and Python
• Strong problem-solving and analytical skills

ACHIEVEMENTS
In my previous role, I successfully led a team project that resulted in a 40% improvement in application performance. I have also contributed to open-source projects and maintained a strong commitment to clean, maintainable code.

I am excited about the opportunity to contribute to your organization's continued success and would welcome the chance to discuss how my skills and enthusiasm can benefit your team.

Thank you for your time and consideration.

Sincerely,
[Your Name]"""
        else:
            document_content = f"""DOCUMENT: {filename}

This is a sample document that demonstrates the Word Editor functionality.

INTRODUCTION
This document contains sample text that would normally be extracted from an uploaded .docx file. In a production environment, the system would use libraries like python-docx to extract the actual content from uploaded documents.

CONTENT SECTIONS

Section 1: Document Overview
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris.

Section 2: Features and Capabilities
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

Section 3: Technical Implementation
Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.

CONCLUSION
This document demonstrates how uploaded content would appear in the Word Editor. Users can apply various formatting options including font changes, alignment, spacing, and more.

---
Document Information:
- Filename: {filename}
- Upload Date: {datetime.now().strftime('%Y-%m-%d')}
- Processing Status: Complete"""

        global doc_id_counter
        new_doc = {
            'id': str(doc_id_counter),
            'filename': filename,
            'content': document_content,
            'uploaded_at': datetime.utcnow(),
            'formatting': DEFAULT_FORMATTING.copy()
        }

        documents.append(new_doc)
        doc_id_counter += 1

        print("Document created successfully:", new_doc['id'])

        return jsonify({
            'success': True,
            'message': 'Document uploaded successfully',
            'documentId': new_doc['id'],
            'content': document_content
        })

    except Exception as e:
        print(f"Upload document error: {e}")
        return jsonify({
            'success': False,
            'message': f'Failed to upload document: {str(e)}'
        }), 500

@word_editor_bp.route('/edit-doc', methods=['POST'])
def edit_doc():
    try:
        data = request.get_json()
        document_id = data.get('documentId')
        formatting = data.get('formatting', {})
        content = data.get('content')

        # Find document
        doc = next((d for d in documents if d['id'] == document_id), None)
        if not doc:
            return jsonify({
                'success': False,
                'message': 'Document not found'
            }), 404

        doc['formatting'] = merge_formatting(doc.get('formatting', DEFAULT_FORMATTING), formatting)
        if isinstance(content, str):
            doc['content'] = content
            print(f"Content updated. New length: {len(content)}")
            print(f"Content preview (first 200 chars): {content[:200]}")

        # In production, apply formatting using python-docx or similar
        print("Applying formatting:", doc['formatting'])
        print(f"Document state after edit - Content length: {len(doc.get('content', ''))}, Formatting keys: {list(doc.get('formatting', {}).keys())}")

        return jsonify({
            'success': True,
            'message': 'Document formatting applied successfully',
            'preview': doc['content']
        })

    except Exception as e:
        print(f"Edit document error: {e}")
        return jsonify({
            'success': False,
            'message': f'Failed to edit document: {str(e)}'
        }), 500

@word_editor_bp.route('/download-doc/<document_id>')
def download_doc(document_id):
    try:
        # Find document
        doc = next((d for d in documents if d['id'] == document_id), None)
        if not doc:
            return jsonify({
                'success': False,
                'message': 'Document not found'
            }), 404

        export_format = request.args.get('format', 'docx').lower()
        original_name = doc['filename']
        
        # Debug: Log what content and formatting we're using
        print(f"Downloading document {document_id}")
        print(f"Content length: {len(doc.get('content', ''))}")
        print(f"Content preview (first 200 chars): {doc.get('content', '')[:200]}")
        print(f"Formatting: {doc.get('formatting', {})}")

        if export_format == 'pdf':
            file_buffer = build_pdf_document(doc)
            mimetype = 'application/pdf'
            download_name = f"{original_name.rsplit('.', 1)[0]}_edited.pdf"
        else:
            file_buffer = build_docx_document(doc)
            mimetype = 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
            download_name = f"{original_name.rsplit('.', 1)[0]}_edited.docx"

        return send_file(
            file_buffer,
            mimetype=mimetype,
            as_attachment=True,
            download_name=download_name
        )

    except Exception as e:
        print(f"Download document error: {e}")
        return jsonify({
            'success': False,
            'message': f'Failed to download document: {str(e)}'
        }), 500

File: StudyAI-main/server/routes/word-editor.ts
import { RequestHandler } from "express";
import { UploadDocResponse, EditDocRequest } from "@shared/api";
import PDFDocument from "pdfkit";
import { Document, Packer, Paragraph, TextRun, AlignmentType, Footer, PageNumber } from "docx";

// Simple in-memory document storage for demo
interface DocumentFormatting {
  fontFamily?: string;
  fontSize?: number;
  fontColor?: string;
  margins?: {
    top: number;
    bottom: number;
    left: number;
    right: number;
  };
  alignment?: 'left' | 'center' | 'right' | 'justify';
  lineSpacing?: number;
  pageNumbers?: boolean;
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
}

interface StoredDocument {
  id: string;
  filename: string;
  content: string;
  formatting?: DocumentFormatting;
  uploadedAt: Date;
}

const documents: StoredDocument[] = [];
let docIdCounter = 1;

export const handleUploadDoc: RequestHandler = (req, res) => {
  try {
    console.log("Upload request received");
    console.log("Content-Type:", req.headers['content-type']);
    console.log("Body:", req.body);

    // For demo purposes, we'll simulate file upload and content extraction
    // In production, use multer middleware to handle actual file uploads

    let filename = 'uploaded_document.docx';
    let documentContent = '';

    // Get filename from JSON request body
    if (req.body && req.body.filename) {
      filename = req.body.filename;
    }

    // Generate realistic document content based on filename
    if (filename.toLowerCase().includes('resume') || filename.toLowerCase().includes('cv')) {
      documentContent = `RESUME

John Doe
Software Engineer
Email: john.doe@email.com
Phone: (555) 123-4567

EXPERIENCE
Senior Software Engineer at TechCorp (2020-Present)
• Developed web applications using React and Node.js
• Led a team of 5 developers
• Improved application performance by 40%

Software Engineer at StartupXYZ (2018-2020)
• Built mobile applications using React Native
• Collaborated with design team on user interfaces

EDUCATION
Bachelor of Computer Science
University of Technology (2014-2018)

SKILLS
• JavaScript, Python, Java
• React, Node.js, MongoDB
• Git, Docker, AWS

ACHIEVEMENTS
• Published 3 technical articles
• Speaker at 2 tech conferences
• AWS Certified Solutions Architect`;
    } else if (filename.toLowerCase().includes('report') || filename.toLowerCase().includes('analysis')) {
      documentContent = `PROJECT REPORT

Title: Analysis of Student Performance in Online Learning

ABSTRACT
This report analyzes the impact of online learning on student performance during the academic year 2023-2024. The study examines various factors affecting student engagement and academic outcomes.

INTRODUCTION
The transition to online learning has significantly changed educational methodologies. This study examines the effectiveness of digital learning platforms and their impact on student achievement.

METHODOLOGY
Data was collected from 500 students across different departments using:
• Online surveys and questionnaires
• Academic performance records
• Engagement metrics from learning platforms
• Focus group discussions

FINDINGS
1. Student engagement decreased by 15% in online formats
2. Technical difficulties affected 30% of students
3. Self-motivated students performed better in online settings
4. Interactive content improved retention by 25%

CONCLUSION
Online learning presents both challenges and opportunities for educational institutions. Proper implementation and support systems are crucial for success.

RECOMMENDATIONS
• Improve technical infrastructure
• Provide training for both students and faculty
• Develop hybrid learning models
• Implement interactive learning tools`;
    } else if (filename.toLowerCase().includes('letter') || filename.toLowerCase().includes('application')) {
      documentContent = `APPLICATION LETTER

[Date: ${new Date().toLocaleDateString()}]

Dear Hiring Manager,

I am writing to express my strong interest in the Software Engineer position at your esteemed organization. With my background in computer science and hands-on experience in web development, I am confident that I would be a valuable addition to your team.

QUALIFICATIONS
• Bachelor's degree in Computer Science
• 3+ years of experience in software development
• Proficiency in React, Node.js, and Python
• Strong problem-solving and analytical skills

ACHIEVEMENTS
In my previous role, I successfully led a team project that resulted in a 40% improvement in application performance. I have also contributed to open-source projects and maintained a strong commitment to clean, maintainable code.

I am excited about the opportunity to contribute to your organization's continued success and would welcome the chance to discuss how my skills and enthusiasm can benefit your team.

Thank you for your time and consideration.

Sincerely,
[Your Name]`;
    } else {
      documentContent = `DOCUMENT: ${filename}

This is a sample document that demonstrates the Word Editor functionality.

INTRODUCTION
This document contains sample text that would normally be extracted from an uploaded .docx file. In a production environment, the system would use libraries like python-docx to extract the actual content from uploaded documents.

CONTENT SECTIONS

Section 1: Document Overview
Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris.

Section 2: Features and Capabilities
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

Section 3: Technical Implementation
Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.

CONCLUSION
This document demonstrates how uploaded content would appear in the Word Editor. Users can apply various formatting options including font changes, alignment, spacing, and more.

---
Document Information:
- Filename: ${filename}
- Upload Date: ${new Date().toLocaleDateString()}
- Processing Status: Complete`;
    }

    const newDoc: StoredDocument = {
      id: docIdCounter.toString(),
      filename,
      content: documentContent,
      formatting: {
        fontFamily: 'Arial',
        fontSize: 12,
        fontColor: '#000000',
        margins: { top: 1, bottom: 1, left: 1, right: 1 },
        alignment: 'left',
        lineSpacing: 1.5,
        pageNumbers: false,
        bold: false,
        italic: false,
        underline: false
      },
      uploadedAt: new Date()
    };

    documents.push(newDoc);
    docIdCounter++;

    console.log("Document created successfully:", newDoc.id);

    const response: UploadDocResponse = {
      success: true,
      message: "Document uploaded successfully",
      documentId: newDoc.id,
      content: documentContent
    };

    res.json(response);
  } catch (error) {
    console.error("Upload document error:", error);
    const response: UploadDocResponse = {
      success: false,
      message: `Failed to upload document: ${error instanceof Error ? error.message : 'Unknown error'}`
    };
    res.status(500).json(response);
  }
};

export const handleEditDoc: RequestHandler = (req, res) => {
  try {
    const { documentId, formatting, content }: EditDocRequest & { content?: string } = req.body;

    // Find document
    const doc = documents.find(d => d.id === documentId);
    if (!doc) {
      return res.status(404).json({
        success: false,
        message: "Document not found"
      });
    }

    // Save formatting if provided
    if (formatting) {
      doc.formatting = {
        ...doc.formatting,
        ...formatting
      };
      console.log("Formatting saved:", doc.formatting);
    }

    // Save content if provided
    if (typeof content === 'string') {
      doc.content = content;
      console.log("Content updated. New length:", content.length);
      console.log("Content preview (first 200 chars):", content.substring(0, 200));
    }

    res.json({
      success: true,
      message: "Document formatting applied successfully",
      preview: doc.content
    });
  } catch (error) {
    console.error("Edit document error:", error);
    res.status(500).json({
      success: false,
      message: "Failed to apply formatting"
    });
  }
};

export const handleDownloadDoc: RequestHandler = (req, res) => {
  try {
    const { documentId } = req.params;
    const { format } = req.query;

    // Find document
    const doc = documents.find(d => d.id === documentId);
    if (!doc) {
      return res.status(404).json({
        success: false,
        message: "Document not found"
      });
    }

    // Debug: Log what we're using
    console.log(`Downloading document ${documentId}`);
    console.log(`Content length: ${doc.content.length}`);
    console.log(`Content preview (first 200 chars): ${doc.content.substring(0, 200)}`);
    console.log(`Formatting:`, doc.formatting);

    const baseFilename = doc.filename.replace('.docx', '').replace('.pdf', '');
    const filename = format === 'pdf' ? `${baseFilename}_edited.pdf` : `${baseFilename}_edited.docx`;
    
    const formatting = doc.formatting || {
      fontFamily: 'Arial',
      fontSize: 12,
      fontColor: '#000000',
      margins: { top: 1, bottom: 1, left: 1, right: 1 },
      alignment: 'left',
      lineSpacing: 1.5,
      pageNumbers: false,
      bold: false,
      italic: false,
      underline: false
    };

    if (format === 'pdf') {
      // Generate PDF using PDFKit with formatting
      const margins = formatting.margins || { top: 1, bottom: 1, left: 1, right: 1 };
      const pdfDoc = new PDFDocument({
        size: 'A4',
        margins: {
          top: margins.top * 72,
          bottom: margins.bottom * 72,
          left: margins.left * 72,
          right: margins.right * 72
        }
      });

      // Set response headers for PDF
      res.setHeader('Content-Type', 'application/pdf');
      res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);

      // Pipe PDF to response
      pdfDoc.pipe(res);

      // Font mapping for PDF (PDFKit uses standard PDF fonts)
      const fontMap: { [key: string]: string } = {
        'Arial': 'Helvetica',
        'Calibri': 'Helvetica',
        'Georgia': 'Times-Roman',
        'Times New Roman': 'Times-Roman'
      };
      
      const fontFamily = formatting.fontFamily || 'Arial';
      const baseFont = fontMap[fontFamily] || 'Helvetica';
      
      // Function to get the correct font variant based on bold/italic
      const getFontVariant = (isBold: boolean, isItalic: boolean, isHeader: boolean = false) => {
        const useBold = isBold || isHeader;
        const useItalic = isItalic;
        
        if (baseFont === 'Times-Roman') {
          if (useBold && useItalic) return 'Times-BoldItalic';
          if (useBold) return 'Times-Bold';
          if (useItalic) return 'Times-Italic';
          return 'Times-Roman';
        } else {
          // Helvetica
          if (useBold && useItalic) return 'Helvetica-BoldOblique';
          if (useBold) return 'Helvetica-Bold';
          if (useItalic) return 'Helvetica-Oblique';
          return 'Helvetica';
        }
      };

      const fontSize = formatting.fontSize || 12;
      const alignment = formatting.alignment || 'left';
      const alignMap: { [key: string]: 'left' | 'center' | 'right' | 'justify' } = {
        'left': 'left',
        'center': 'center',
        'right': 'right',
        'justify': 'left' // PDFKit doesn't support justify, use left
      };

      // Track pages for page numbering
      let pageNumber = 0;
      
      // Add page numbers if enabled - add to each page as it's created
      if (formatting.pageNumbers) {
        // Add page number to first page immediately
        pageNumber = 1;
        const firstPage = pdfDoc.page;
        const savedY = pdfDoc.y;
        const savedX = pdfDoc.x;
        
        pdfDoc.fontSize(10)
              .font('Helvetica')
              .fillColor('#666666')
              .text(`Page ${pageNumber}`, 0, firstPage.height - 40, {
                width: firstPage.width,
                align: 'center'
              });
        
        // Restore cursor to start writing content
        pdfDoc.x = savedX;
        pdfDoc.y = savedY;
        
        // Add page numbers to subsequent pages
        pdfDoc.on('pageAdded', () => {
          pageNumber++;
          const currentPageIndex = pageNumber - 1;
          
          // Switch to the newly added page
          pdfDoc.switchToPage(currentPageIndex);
          
          // Save current cursor position
          const savedY = pdfDoc.y;
          const savedX = pdfDoc.x;
          
          // Add page number at the bottom center
          const page = pdfDoc.page;
          pdfDoc.fontSize(10)
                .font('Helvetica')
                .fillColor('#666666')
                .text(`Page ${pageNumber}`, 0, page.height - 40, {
                  width: page.width,
                  align: 'center'
                });
          
          // Restore cursor position
          pdfDoc.x = savedX;
          pdfDoc.y = savedY;
        });
      }

      // Add main content with formatting
      const contentLines = doc.content.split('\n');
      contentLines.forEach(line => {
        if (line.trim()) {
          const isHeader = line.match(/^[A-Z\s]+$/) && line.length < 50;
          const lineFontSize = isHeader ? fontSize + 2 : fontSize;
          const lineFont = getFontVariant(formatting.bold || false, formatting.italic || false, isHeader);
          
          pdfDoc.fontSize(lineFontSize)
                .font(lineFont)
                .fillColor(formatting.fontColor || '#000000')
                .text(line.trim(), { 
                  align: alignMap[alignment] || 'left',
                  underline: formatting.underline || false
                });
          pdfDoc.moveDown((formatting.lineSpacing || 1.5) * 0.5);
        } else {
          pdfDoc.moveDown((formatting.lineSpacing || 1.5) * 0.3);
        }
      });

      // Finalize PDF
      pdfDoc.end();

    } else {
      // Generate DOCX using docx library with formatting
      const margins = formatting.margins || { top: 1, bottom: 1, left: 1, right: 1 };
      const alignmentMap: { [key: string]: AlignmentType } = {
        'left': AlignmentType.LEFT,
        'center': AlignmentType.CENTER,
        'right': AlignmentType.RIGHT,
        'justify': AlignmentType.JUSTIFIED
      };

      // Build footer with page numbers if enabled
      const footer = formatting.pageNumbers ? new Footer({
        children: [
          new Paragraph({
            children: [
              new TextRun({
                children: ["Page ", PageNumber.CURRENT, " of ", PageNumber.TOTAL_PAGES],
                size: 20, // 10pt in half-points
                color: '666666'
              })
            ],
            alignment: AlignmentType.CENTER
          })
        ]
      }) : undefined;

      const docxDoc = new Document({
        sections: [{
          properties: {
            page: {
              margin: {
                top: margins.top * 1440, // Convert inches to twips (1 inch = 1440 twips)
                bottom: margins.bottom * 1440,
                left: margins.left * 1440,
                right: margins.right * 1440
              }
            }
          },
          headers: {},
          footers: {
            default: footer
          },
          children: [
            // Main content with formatting
            ...doc.content.split('\n').map(line => {
              const trimmedLine = line.trim();
              if (!trimmedLine) {
                return new Paragraph({ children: [new TextRun({ text: '' })] });
              }
              
              const isHeader = trimmedLine.match(/^[A-Z\s]+$/) && trimmedLine.length < 50;
              const fontSize = isHeader ? (formatting.fontSize || 12) + 2 : (formatting.fontSize || 12);
              
              // Parse color (remove # if present)
              const color = (formatting.fontColor || '#000000').replace('#', '');
              
              // Ensure font family is one of the supported fonts
              const fontFamily = formatting.fontFamily || 'Arial';
              const validFonts = ['Arial', 'Times New Roman', 'Georgia', 'Calibri'];
              const selectedFont = validFonts.includes(fontFamily) ? fontFamily : 'Arial';
              
              return new Paragraph({
                children: [
                  new TextRun({
                    text: trimmedLine,
                    font: selectedFont,
                    size: fontSize * 2, // docx uses half-points
                    color: color,
                    bold: isHeader || formatting.bold || false,
                    italics: formatting.italic || false,
                    underline: formatting.underline ? {} : undefined
                  })
                ],
                alignment: alignmentMap[formatting.alignment || 'left'] || AlignmentType.LEFT,
                spacing: {
                  line: (formatting.lineSpacing || 1.5) * 240, // Convert to twips
                  lineRule: 'auto'
                }
              });
            })
          ]
        }]
      });

      // Set response headers for DOCX
      res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document');
      res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);

      // Generate and send DOCX buffer
      Packer.toBuffer(docxDoc).then(buffer => {
        res.send(buffer);
      }).catch(error => {
        console.error("DOCX generation error:", error);
        res.status(500).json({
          success: false,
          message: "Failed to generate DOCX file"
        });
      });
    }
  } catch (error) {
    console.error("Download document error:", error);
    res.status(500).json({
      success: false,
      message: "Failed to download document"
    });
  }
};
